aws/retry/middleware.go

```diff
@@ -87,61 +83,103 @@ func (r Attempt) HandleFinalize(ctx context.Context, in smithymiddle.FinalizeInp
 			AttemptClockSkew: attemptClockSkew,
 		})
 
-		if attemptNum > 1 {
-			if rewindable, ok := in.Request.(interface{ RewindStream() error }); ok {
-				if err := rewindable.RewindStream(); err != nil {
-					return out, metadata, fmt.Errorf("failed to rewind transport stream for retry, %w", err)
-				}
-			}
+		var attemptResult AttemptResult
 
-			r.logf(logger, logging.Debug, "retrying request %s/%s, attempt %d", service, operation, attemptNum)
-		}
+		out, attemptResult, err = r.handleAttempt(attemptCtx, attemptInput, next)
 
-		out, metadata, reqErr := next.HandleFinalize(attemptCtx, attemptInput)
+		responseMetadata := awsmiddle.GetResponseMetadata(attemptResult.ResponseMetadata)
+		attemptClockSkew = responseMetadata.AttemptSkew
+		shouldRetry = attemptResult.Retried
 
-		if releaseError := relRetryToken(reqErr); releaseError != nil && reqErr != nil {
-			return out, metadata, fmt.Errorf("failed to release token after request error, %v", reqErr)
-		}
+		// add attempt metadata to list of all attempt metadata
+		attemptResults.Results = append(attemptResults.Results, attemptResult)
 
-		if reqErr == nil {
-			return out, metadata, nil
+		if !shouldRetry {
+			break
 		}
+	}
 
-		retryable := r.retryer.IsErrorRetryable(reqErr)
-		if !retryable {
-			r.logf(logger, logging.Debug, "request failed with unretryable error %v", reqErr)
-			return out, metadata, reqErr
-		}
+	addAttemptResults(&metadata, attemptResults)
+	return out, metadata, err
+}
+
+// handleAttempt handles an individual request attempt.
+func (r Attempt) handleAttempt(ctx context.Context, in smithymiddle.FinalizeInput, next smithymiddle.FinalizeHandler) (
+	out smithymiddle.FinalizeOutput, attemptResult AttemptResult, err error,
+) {
+	defer func() {
+		attemptResult.Err = err
+		attemptResult.Response = out.Result
```

Comment:
`out.Result` will be the deserialized output shape of the operation call, not the Raw response message. In order to get things like status code and headers RawResponse is needed.

I think another middleware is needed to add `RawResponse` to the `middleware.Metadata` as the most `before` in the `Deserialize` step.

Reply:
If this pattern is used the explict `Response` member isn't needed and instead could have a method on `AttemptResult` that attempts to retrieve the RawResponse from the nested `ResponseMetadata.